finetrainers/trainer.py

```diff
@@ -0,0 +1,542 @@
+        self.state.learning_rate = self.args.lr
+        if self.args.scale_lr:
+            self.state.learning_rate = self.state.learning_rate * self.args.gradient_accumulation_steps * self.args.batch_size * self.state.accelerator.num_processes
+        
+        transformer_lora_parameters = list(filter(lambda p: p.requires_grad, self.transformer.parameters()))
+        transformer_parameters_with_lr = {
+            "params": transformer_lora_parameters,
+            "lr": self.state.learning_rate,
+        }
+        params_to_optimize = [transformer_parameters_with_lr]
+        self.state.num_trainable_parameters = sum(p.numel() for p in transformer_lora_parameters)
+
+        # TODO(aryan): add deepspeed support
+        optimizer = get_optimizer(
+            params_to_optimize=params_to_optimize,
+            optimizer_name=self.args.optimizer,
+            learning_rate=self.args.lr,
+            beta1=self.args.beta1,
+            beta2=self.args.beta2,
+            beta3=self.args.beta3,
+            epsilon=self.args.epsilon,
+            weight_decay=self.args.weight_decay,
+        )
+
+        num_update_steps_per_epoch = math.ceil(len(self.dataloader) / self.args.gradient_accumulation_steps)
+        if self.state.train_steps is None:
+            self.state.train_steps = self.state.train_epochs * num_update_steps_per_epoch
+            self.state.overwrote_max_train_steps = True
+
+        lr_scheduler = get_scheduler(
+            name=self.args.lr_scheduler,
+            optimizer=optimizer,
+            num_warmup_steps=self.args.lr_warmup_steps * self.state.accelerator.num_processes,
+            num_training_steps=self.state.train_steps * self.state.accelerator.num_processes,
+            num_cycles=self.args.lr_num_cycles,
+            power=self.args.lr_power,
+        )
+
+        self.optimizer = optimizer
+        self.lr_scheduler = lr_scheduler
+    
+    def prepare_for_training(self) -> None:
+        self.transformer, self.optimizer, self.dataloader, self.lr_scheduler = self.state.accelerator.prepare(
+            self.transformer, self.optimizer, self.dataloader, self.lr_scheduler
+        )
+
+        # We need to recalculate our total training steps as the size of the training dataloader may have changed.
+        num_update_steps_per_epoch = math.ceil(len(self.dataloader) / self.args.gradient_accumulation_steps)
+        if self.state.overwrote_max_train_steps:
+            self.state.train_steps = self.state.train_epochs * num_update_steps_per_epoch
+        # Afterwards we recalculate our number of training epochs
+        self.state.train_epochs = math.ceil(self.state.train_steps / num_update_steps_per_epoch)
+    
+    def prepare_trackers(self) -> None:
+        logger.info("Initializing trackers")
+
+        tracker_name = self.args.tracker_name or "finetrainers-experiment"
+        self.state.accelerator.init_trackers(tracker_name, config=self.args.to_dict())
+    
+    def train(self) -> None:
```

Comment:
I think this could be factored into include methods like the following:

* `prepare_inputs_for_loss()`
* `compute_loss()`

Both of these differ in Mochi-1 from the standard ones like Cog.

Reply:
That sounds good to me. For now, I will keep this as-is and looking into how best to abstract these details when I get to Mochi, Cog, and Hunyuan. I think we could nicely design it around most common training strategies